src/post/solution.rs

```diff
@@ -0,0 +1,120 @@
+    pub id: SolutionId,
+    pub name: String,
+    pub description: String,
+    pub author_id: AccountId,
+    #[serde(with = "u64_dec_format")]
+    pub timestamp: Timestamp,
+    pub status: PostStatus,
+    pub likes: HashSet<Like>,
+    pub comments: Vec<CommentId>,
+
+    // Specific fields
+    #[serde(with = "u64_dec_format")]
+    pub idea_id: u64,
+    pub attestations: Vec<AttestationId>,
+    pub sponsorships: Vec<SponsorshipId>,
+}
+
+#[derive(BorshDeserialize, BorshSerialize, Serialize, Deserialize, Clone)]
+#[serde(crate = "near_sdk::serde")]
+pub struct SolutionV1 {
+    pub name: String,
+    pub description: String,
+}
+
+#[derive(BorshDeserialize, BorshSerialize, Serialize, Deserialize, Clone)]
+#[serde(crate = "near_sdk::serde")]
+pub struct SolutionV2 {
+    pub name: String,
+    pub description: String,
+    pub requested_sponsor: Option<AccountId>,
+    #[serde(with = "u128_dec_format")]
+    pub requested_sponsorship_amount: Balance,
+    pub sponsorship_token: Option<SponsorshipToken>,
+}
+
+#[derive(BorshDeserialize, BorshSerialize, Serialize, Deserialize, Clone)]
+#[serde(crate = "near_sdk::serde")]
+#[serde(tag = "solution_version")]
+pub enum VersionedSolution {
+    V0(SolutionV0),
+    V1(SolutionV1),
+    V2(SolutionV2),
+}
+
+impl VersionedSolution {
+    pub fn latest_version(self) -> SolutionV2 {
+        self.into()
+    }
+
+    pub fn validate(&self) {
+        match self {
+            VersionedSolution::V2(solution) => {
+                if solution.requested_sponsorship_amount > 0
+                    && (solution.sponsorship_token.is_none()
+                        || solution.requested_sponsor.is_none())
+                {
+                    panic!(
+                        "Solution that requires funding must specify sponsorship token and sponsor"
+                    )
+                }
+            }
+
+            _ => unimplemented!(),
+        }
+    }
+}
+
+impl From<VersionedSolution> for SolutionV0 {
+    fn from(solution: VersionedSolution) -> Self {
+        match solution {
+            VersionedSolution::V0(v0) => v0,
+            _ => unimplemented!(),
+        }
+    }
+}
+
+impl From<VersionedSolution> for SolutionV1 {
+    fn from(solution: VersionedSolution) -> Self {
+        match solution {
+            VersionedSolution::V1(v1) => v1,
+            _ => unimplemented!(),
+        }
+    }
+}
+
+impl From<SolutionV0> for VersionedSolution {
+    fn from(solution: SolutionV0) -> Self {
+        VersionedSolution::V0(solution)
+    }
+}
+
+impl From<VersionedSolution> for SolutionV2 {
+    fn from(solution: VersionedSolution) -> Self {
+        match solution {
+            VersionedSolution::V2(v2) => v2,
+
+            VersionedSolution::V1(v1) => SolutionV2 {
+                name: v1.name,
+                description: v1.description,
+                requested_sponsor: None,
+                requested_sponsorship_amount: 0,
+                sponsorship_token: None,
```

Comment:
I feel this conversion is worse than not doing the conversion. Can we store both V1 and V2 in the state, and only when user edit it, it will store V2 in contract?

Reply:
@ailisp yes, this is how it's eventually going to work, since we're not introducing a migration. 0 and None are needed only for solutions that don't have funding requests